File: sol/725/725.cpp
struct Solution {
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        ListNode *ptr = head;
        int l = 0;
        while (ptr != nullptr){
            ++l;
            ptr = ptr->next;
        }
        int q = l / k, r = l % k;
        vector<ListNode *> res(k, nullptr);
        for (int i = 0; i < r; ++i){
            res[i] = head;
            ptr = head;
            head = head->next;
            for (int j = 0; j < q; ++j){
                ptr->next = head;
                ptr = head;
                head = head->next;
            }
            ptr->next = nullptr;
        }
        for (int i = r; i < k; ++i){
            if (head == nullptr){
                continue;
            }
            res[i] = head;
            ptr = head;
            head = head->next;
            for (int j = 0; j < q - 1; ++j){
                ptr->next = head;
                ptr = head;
                head = head->next;
            }
            ptr->next = nullptr;
        }
        return res;
    }
};

File: sol/725/725.py
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        ptr = head
        l = 0
        while ptr is not None:
            ptr = ptr.next
            l += 1
        q, r = divmod(l, k)
        res = [None for _ in range(k)]
        for i in range(r):
            if head is None:
                continue
            res[i] = head
            ptr = head
            head = head.next
            for _ in range(q):
                ptr.next = head
                ptr = head
                head = head.next
            ptr.next = None
        for i in range(r, k):
            if head is None:
                continue
            res[i] = head
            ptr = head
            head = head.next
            for _ in range(q - 1):
                ptr.next = head
                ptr = head
                head = head.next
            ptr.next = None
        return res
